src/controllers/hid/hidcontroller.cpp

```diff
@@ -124,196 +145,72 @@ int HidController::close() {
-
-    qCDebug(m_logInput) << bytesRead
-                        << "bytes received by hid_get_input_report" << getName()
-                        << "serial #" << m_deviceInfo.serialNumber()
-                        << "(including one byte for the report ID:"
-                        << QString::number(static_cast<quint8>(reportID), 16)
-                                   .toUpper()
-                                   .rightJustified(2, QChar('0'))
-                        << ")";
-
-    if (bytesRead <= kReportIdSize) {
-        // -1 is the only error value according to hidapi documentation.
-        // Otherwise minimum possible value is 1, because 1 byte is for the reportID,
-        // the smallest report with data is therefore 2 bytes.
-        DEBUG_ASSERT(bytesRead <= kReportIdSize);
-        return QByteArray();
-    }
+        qCInfo(m_logBase) << "Waiting on HID IO thread to send the last remaining OutputReports";
+        VERIFY_OR_DEBUG_ASSERT(m_pHidIoThread->waitUntilRunLoopIsStopped(50)) {
+            qWarning() << "Sending the last remaining OutputReports "
+                          "reached timeout!";
 
-    return QByteArray::fromRawData(
-            reinterpret_cast<char*>(m_pPollData[m_pollingBufferIndex]), bytesRead);
-}
+            qCInfo(m_logBase) << "Enforce stop of HID IO thread run loop!";
+            m_pHidIoThread->setThreadState(HidIoThreadState::StopRequested);
 
-bool HidController::poll() {
-    Trace hidRead("HidController poll");
-
-    // This loop risks becoming a high priority endless loop in case processing
-    // the mapping JS code takes longer than the controller polling rate.
-    // This could stall other low priority tasks.
-    // There is no safety net for this because it has not been demonstrated to be
-    // a problem in practice.
-    while (true) {
-        int bytesRead = hid_read(m_pHidDevice, m_pPollData[m_pollingBufferIndex], kBufferSize);
-        if (bytesRead < 0) {
-            // -1 is the only error value according to hidapi documentation.
-            DEBUG_ASSERT(bytesRead == -1);
-            return false;
-        } else if (bytesRead == 0) {
-            // No packet was available to be read
-            return true;
+            VERIFY_OR_DEBUG_ASSERT(m_pHidIoThread->waitUntilRunLoopIsStopped(100)) {
+                qWarning() << "Stopping run loop reached timeout!";
+            }
         }
-        processInputReport(bytesRead);
+
+        // After completion of all HID communication deconstruct m_pHidIoThread
+        m_pHidIoThread.reset();
     }
-}
 
-bool HidController::isPolling() const {
-    return isOpen();
+    // Close device
+    setOpen(false);
+    qCInfo(m_logBase) << "Device closed";
+    return 0;
 }
 
 void HidController::sendReport(QList<int> data, unsigned int length, unsigned int reportID) {
     Q_UNUSED(length);
     QByteArray temp;
+    temp.reserve(data.size());
     foreach (int datum, data) {
```

Comment:
Lets replace this custom Qt macro with a range-base for loop. It requires that the parameter type is changed from `QList<int>` to `const QList<int>&` if we want to avoid using `qAsConst()`.

Reply:
...then `temp.reserve(...)` becomes obsolete.